File: problems/847_shortest_path_visiting_all_nodes/solution.cpp
#include <vector>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <algorithm>
#include <climits>
#include <queue>
#include <stack>
#include <cmath>
#include <numeric>
#include <set>
#include <map>
using namespace std;

class Solution {
public:
    int shortestPathLength(vector<vector<int>>& graph) {
        int n = graph.size(), target = (1 << n) - 1;
        queue<tuple<int,int,int>> q;
        set<pair<int,int>> visited;
        for (int i = 0; i < n; i++) {
            q.push({i, 1 << i, 0});
            visited.insert({i, 1 << i});
        }
        while (!q.empty()) {
            auto [node, mask, dist] = q.front(); q.pop();
            if (mask == target) return dist;
            for (int nei : graph[node]) {
                int newMask = mask | (1 << nei);
                if (!visited.count({nei, newMask})) {
                    visited.insert({nei, newMask});
                    q.push({nei, newMask, dist + 1});
                }
            }
        }
        return -1;
    }
};

File: problems/847_shortest_path_visiting_all_nodes/solution.py
from typing import List


class Solution:
    def shortestPathLength(self, graph: List[List[int]]) -> int:
        n = len(graph)
        target = (1 << n) - 1
        queue = deque()
        visited = set()
        for i in range(n):
            state = (i, 1 << i)
            queue.append((i, 1 << i, 0))
            visited.add(state)
        while queue:
            node, mask, dist = queue.popleft()
            if mask == target:
                return dist
            for nei in graph[node]:
                new_mask = mask | (1 << nei)
                if (nei, new_mask) not in visited:
                    visited.add((nei, new_mask))
                    queue.append((nei, new_mask, dist + 1))
        return -1
